File: src/main/scala/io/jobial/scase/core/RequestResponseClient.scala
package io.jobial.scase.core

import cats.effect.IO

import scala.annotation.implicitNotFound
import scala.concurrent.duration.Duration
import scala.reflect.ClassTag


trait RequestResult[RESPONSE] {

  def response: IO[MessageReceiveResult[RESPONSE]]

  def commit: IO[_]
}

case class SendRequestContext(
  requestTimeout: Duration,
  attributes: Map[String, String] = Map()
)

@implicitNotFound("No mapping found from request type ${REQUEST} to response type ${RESPONSE}")
trait RequestResponseMapping[REQUEST, RESPONSE]

trait RequestResponseClient[REQ, RESP] {

  def sendRequest[REQUEST <: REQ, RESPONSE <: RESP](request: REQUEST)
    (implicit requestResponseMapping: RequestResponseMapping[REQUEST, RESPONSE], sendRequestContext: SendRequestContext): RequestResult[RESPONSE]

  def ?[REQUEST <: REQ, RESPONSE <: RESP : ClassTag](request: REQUEST)
    (implicit requestResponseMapping: RequestResponseMapping[REQUEST, RESPONSE], sendRequestContext: SendRequestContext) =
    sendRequest(request).response.map(_.message)
}

File: scase-core/src/main/java/io/jobial/scase/core/javadsl/RequestResponseClient.java
package io.jobial.scase.core.javadsl;

import cats.effect.IO;
import io.jobial.scase.core.RequestResponseMapping;
import io.jobial.scase.core.RequestResponseResult;
import io.jobial.scase.core.RequestTimeout;

import java.util.concurrent.CompletableFuture;

import static io.jobial.scase.core.javadsl.JavaUtils.ioToCompletableFuture;

public class RequestResponseClient<REQ, RESP> {

    private io.jobial.scase.core.RequestResponseClient<IO, REQ, RESP> client;

    public RequestResponseClient(io.jobial.scase.core.RequestResponseClient<IO, REQ, RESP> client) {
        this.client = client;
    }

    private RequestResponseMapping<REQ, RESP> requestResponseMapping = new RequestResponseMapping<REQ, RESP>() {
    };

    public CompletableFuture<RESP> sendRequest(REQ request, SendRequestContext sendRequestContext) throws RequestTimeout {
        return ioToCompletableFuture(client.sendRequestWithResponseMapping(request, requestResponseMapping, sendRequestContext.getContext()).flatMap(
                JavaUtils.<RequestResponseResult<IO, REQ, RESP>, IO<RESP>>javaFunctionToScala(r -> r.response().message())
        ));
    }

    public CompletableFuture<RESP> sendRequest(REQ request) throws RequestTimeout {
        return sendRequest(request, new SendRequestContext());
    }

    public CompletableFuture<RESP> sendRequestWithFullResult(REQ request) throws RequestTimeout {
        return sendRequest(request, new SendRequestContext());
    }

    public CompletableFuture<?> stop() {
        return ioToCompletableFuture(client.stop());
    }
}
